src/main/ruby/core/truffle/interop.rb

```diff
@@ -24,17 +24,73 @@ def self.import_method(name)
     def self.export_method(name)
       export(name.to_s, Object.method(name.to_sym))
     end
+    
+    def self.object_has_keys(object)
+      case object
+      when NilClass, TrueClass, FalseClass, Fixnum, Bignum, Float, Symbol
+        false
+      else
+        true
+      end
+    end
 
-    def self.object_keys(object)
+    def self.object_keys(object, internal)
       if object.is_a?(Hash)
-        object.keys.map(&:to_s)
+        keys = object.keys.map(&:to_s)
       else
-        object.instance_variables.map { |ivar|
+        keys = object.instance_variables.map { |ivar|
           ivar = ivar.to_s
           ivar = ivar[1..-1] if ivar.start_with?('@')
           ivar
         }
       end
+      if internal
+        object.instance_variables.each do |ivar|
+          ivar = ivar.to_s
+          if ivar.start_with?('@')
+            keys << ivar
+          end
+        end
+      end
+      keys
```

Comment:
It's a bit weird to return twice the ivars if internal is true.
Maybe with internal==false we should only return ivars with a `attr_reader` or if there is a method of the same name?
Sounds tricky and fragile though so the current solution might be best.

Reply:
(non-internal case) If `KEYS` is used to send `READ` later then we should probably only return ivars which have a getter method of same name, otherwise `READ` will fail.

Reply:
These are questions about how to change the semantics of `KEYS` in general, so not right for this PR. We need someone to be using `KEYS`, `KEY_INFO` etc to tell us how it should be refined.